thunder/distributed/__init__.py

```diff
@@ -349,6 +350,162 @@ def f(tensor: TensorProxy) -> str:
     return f
 
 
+def fsdp_transform_module(
+    thunder_model: "thunder.ThunderModule",
+    *,
+    device: torch.device | None = None,
+    broadcast_from: int | None = None,
+    sharding_strategy: FSDPType = FSDPType.ZERO2,
+    bucketing_strategy: FSDPBucketingStrategy = FSDPBucketingStrategy.NONE,
+) -> "thunder.ThunderModule":
+    import thunder
+
+    cd = thunder.compile_data(thunder_model)
+    cd.use_fsdp = True
+
+    process_group = tdist.distributed_c10d._get_default_group()
+    utils.check(process_group is not None, lambda: "The default process group is None")
+    global_rank = tdist.get_rank(group=process_group)
+    world_size = tdist.get_world_size(group=process_group)
+    if device is None:
+        local_rank = int(os.environ["LOCAL_RANK"])
+        device = torch.device("cuda", local_rank)
+
+    def prologue_and_compute_transform(prologue_trace, computation_trace, epilogue_trace, **kwargs):
+        import thunder
+
+        prologue_producers, prologue_consumers = thunder.core.utils.producers_and_consumers(prologue_trace)
+        computation_producers, computation_consumers = thunder.core.utils.producers_and_consumers(computation_trace)
+
+        modules_and_thunder_modules = [
+            (bsym.args[0], bsym.output)
+            for bsym in prologue_trace.bound_symbols
+            if bsym.sym is thunder.prims.unpack_thunder_module
+        ]
+
+        if (
+            len(modules_and_thunder_modules) != 1
+            or prologue_producers[modules_and_thunder_modules[0][0]].sym is not thunder.prims.unpack_function_obj
```

Comment:
could this get more readable?
The hardcoded [0][0] feels not straightforward to me

Reply:
Good call, thank you. I'll move the destructuring up and do the checking afterwards.